# Extreme Learning Machines 

Nous entraînerons le réseau à classer les chiffres manuscrits à l'aide de l'ensemble de données MNIST.

L'importation des packages necessaires.

In [42]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
from scipy.linalg import pinv2
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

On doit telecharger la dataset pour entrainer notre ELM et tester le modele

In [43]:
input_length = 28**2
num_classes = 10

# Load MNIST Dataset 
(x_train, y_train), (x_test, y_test) = mnist.load_data() 
# Process images into input vectors 
# each mnist image is a 28x28 picture with value ranges between 0 and 255 
x_train = x_train.astype(np.float32) / 255. 
x_train = x_train.reshape(-1, input_length) 
x_test = x_test.astype(np.float32) / 255. 
x_test = x_test.reshape(-1, input_length) 
# converts [1,2] into [[0,1,0], [0,0,1]] 
y_train = to_categorical(y_train, num_classes).astype(np.float32) 
y_test = to_categorical(y_test, num_classes).astype(np.float32)

In [44]:
x_test.shape

(10000, 784)

Pour initialiser notre reseaux, nous sommes besoin d'identifier:
1. La taille de la couche d'entrée, qui correspond au nombre des valeurs d'entrée
2. Le nombre des noeuds cachées
3. L'entrée des poids cachées
4. Fonction d'activation de la couche cachée

La taille de la couche d'entrée fait référence au nombre de valeurs en entrée du jeu de données.

In [45]:
input_size = x_train.shape[1]

Initialisons le nombre de neurones cachés à 1000.

In [46]:
hidden_size = 1000

Ensuite, nous devons initialiser les poids d'entrée et les biais tirés au hasard d'une distribution gaussienne.

In [47]:
input_weights = np.random.normal(size=[input_size,hidden_size])
biases = np.random.normal(size=[hidden_size])

Fonction d'activation de la couche cachée (Nous utiliserons une unité linéaire rectifiée (ReLU) comme fonction d'activation de couche cachée)

In [48]:
def relu(x):
    return np.maximum(x, 0, x)

Créons une fonction pour calculer notre vecteur H.

In [51]:
def hidden_nodes(X):
    G = np.dot(X, input_weights)
    G = G + biases
    H = relu(G)
    return H

calculer notre β(output_weights)

In [53]:
output_weights = np.dot(pinv2(hidden_nodes(x_train)), y_train)

Pour s'assurer que notre modèle produit un bon résultat, nous devons d'abord le tester. Créons une fonction pour gérer les tests.

In [54]:
def predict(X):
    out = hidden_nodes(X)
    out = np.dot(out, output_weights)
    return out

In [55]:
prediction = predict(x_test)
correct = 0
total = x_test.shape[0]


for i in range(total):
    predicted = np.argmax(prediction[i])
    actual = np.argmax(y_test[i])
    correct += 1 if predicted == actual else 0
accuracy = correct/total
print('Accuracy for ', hidden_size, ' hidden nodes: ', accuracy)

Accuracy for  1000  hidden nodes:  0.9421


La précision du modèle est de 0.9421, ce qui est déjà un bon résultat compte tenu du fait que nous n'avons qu'une seule couche cachée avec 1000 nœuds cachés et un réglage non itératif pour l'apprentissage, ce qui le rend plus rapide que toutes les techniques basées sur le gradient.